File: apps/web/src/lib/testScenarios.ts
export type TestScenarioMode =
  | "chat"
  | "generate"
  | "interactive_train"
  | "interactive_watch"
  | "interactive_play"
  | "interactive_coach"
  | "interactive_arena";

export interface TestScenario {
  key: string;
  label: string;
  mode: TestScenarioMode;
  placeholder: string;
  hint: string;
}

/** Mirrors `apps/api/brainspa_api/test_scenarios.py` when API is offline or stale. */
export const FALLBACK_SCENARIOS: Record<string, TestScenario[]> = {
  snake_policy: [
    {
      key: "autonomous-train",
      label: "AUTONOMOUS TRAIN",
      mode: "interactive_train",
      placeholder: "",
      hint: "Six boards train in parallel.",
    },
    {
      key: "autonomous-watch",
      label: "AUTONOMOUS WATCH",
      mode: "interactive_watch",
      placeholder: "",
      hint: "Policy plays solo — pick speed.",
    },
    {
      key: "human-play",
      label: "HUMAN PLAY",
      mode: "interactive_play",
      placeholder: "",
      hint: "You control the snake.",
    },
    {
      key: "coach-replay",
      label: "COACH REPLAY",
      mode: "interactive_coach",
      placeholder: "",
      hint: "Step through a saved game.",
    },
    {
      key: "human-vs-ai",
      label: "HUMAN VS AI",
      mode: "interactive_arena",
      placeholder: "",
      hint: "You vs policy on one board.",
    },
    {
      key: "dual-arena",
      label: "DUAL ARENA",
      mode: "interactive_arena",
      placeholder: "",
      hint: "Two policies head to head.",
    },
  ],
};

export function fallbackScenarios(modelKey: string): TestScenario[] {
  return FALLBACK_SCENARIOS[modelKey] ?? [];
}

/** Prefer API scenarios only when they match this model's registry harness. */
export function mergeTestScenarios(modelKey: string, apiScenarios: TestScenario[]): TestScenario[] {
  const expected = fallbackScenarios(modelKey);
  if (!expected.length) {
    return apiScenarios;
  }
  if (!apiScenarios.length) {
    return expected;
  }
  const expectedKeys = new Set(expected.map((scenario) => scenario.key));
  const overlap = apiScenarios.filter((scenario) => expectedKeys.has(scenario.key)).length;
  const minimum = Math.min(2, expectedKeys.size);
  if (overlap >= minimum) {
    return apiScenarios;
  }
  return expected;
}

export { testModelPath as modelPath, testScenarioPath as scenarioPath } from "@/lib/testRoutes";

/** User-facing tuned model name (not registry key). */
const MODEL_DISPLAY_NAMES: Record<string, string> = {
  snake_policy: "Snake Policy",
};

export function modelDisplayName(modelKey: string, apiLabel?: string | null): string {
  if (apiLabel && apiLabel !== "Persona Small") {
    return apiLabel;
  }
  return MODEL_DISPLAY_NAMES[modelKey] ?? apiLabel ?? modelKey;
}

File: apps/api/brainspa_api/test_scenarios.py
from __future__ import annotations

from pydantic import BaseModel, Field


class TestScenarioPublic(BaseModel):
    key: str
    label: str
    mode: str = Field(description="chat | generate | interactive_*")
    placeholder: str = ""
    hint: str = ""


SNAKE_SCENARIOS = [
    TestScenarioPublic(
        key="autonomous-train",
        label="AUTONOMOUS TRAIN",
        mode="interactive_train",
        hint="Six boards train in parallel.",
    ),
    TestScenarioPublic(
        key="autonomous-watch",
        label="AUTONOMOUS WATCH",
        mode="interactive_watch",
        hint="Policy plays solo — pick speed.",
    ),
    TestScenarioPublic(
        key="human-play",
        label="HUMAN PLAY",
        mode="interactive_play",
        hint="You control the snake.",
    ),
    TestScenarioPublic(
        key="coach-replay",
        label="COACH REPLAY",
        mode="interactive_coach",
        hint="Step through a saved game.",
    ),
    TestScenarioPublic(
        key="human-vs-ai",
        label="HUMAN VS AI",
        mode="interactive_arena",
        hint="You vs policy on one board.",
    ),
    TestScenarioPublic(
        key="dual-arena",
        label="DUAL ARENA",
        mode="interactive_arena",
        hint="Two policies head to head.",
    ),
]

SCENARIOS_BY_MODEL: dict[str, list[TestScenarioPublic]] = {"snake_policy": SNAKE_SCENARIOS}

MODEL_SLUG_TO_KEY: dict[str, str] = {"snake": "snake_policy"}


def resolve_test_model_key(slug_or_key: str) -> str | None:
    if slug_or_key in MODEL_SLUG_TO_KEY:
        return MODEL_SLUG_TO_KEY[slug_or_key]
    if slug_or_key in SCENARIOS_BY_MODEL:
        return slug_or_key
    return None


def list_test_scenarios(slug_or_key: str) -> list[TestScenarioPublic]:
    model_key = resolve_test_model_key(slug_or_key)
    if model_key is None:
        return []
    return SCENARIOS_BY_MODEL[model_key]


def scenario_generation_text(scenario_key: str, user_text: str) -> str:
    text = user_text.strip()
    return text
